File: frameworks/playwright-ts/pages/ProductsPage.ts
import { Locator, Page, expect } from "@playwright/test";
import { BasePage } from "./BasePage";

export class ProductsPage extends BasePage {
  readonly productCards: Locator = this.page.locator(".card.product");

  constructor(page: Page) {
    super(page);
  }

  async open() {
    await this.goto("/products");
  }

  cardFor(name: string): Locator {
    return this.productCards.filter({ hasText: name });
  }

  async addToCart(productName: string) {
    await this.cardFor(productName).locator(".add-to-cart").click();
  }

  async expectProductCount(count: number) {
    await expect(this.productCards).toHaveCount(count);
  }

  async flashMessage() {
    return this.page.locator("#flash");
  }
}

File: frameworks/playwright-dotnet/Pages/ProductsPage.cs
using Microsoft.Playwright;

namespace QARonin.PlaywrightDotnet.Pages;

public class ProductsPage
{
    private readonly IPage _page;
    public ProductsPage(IPage page) => _page = page;

    public ILocator ProductCards => _page.Locator(".card.product");
    public ILocator CartCount => _page.Locator("#cart-count");
    public ILocator UserBadge => _page.Locator("#user-badge");

    public async Task OpenAsync(string baseUrl) => await _page.GotoAsync($"{baseUrl}/products");
}
